### ISSUE TITLE
package runtime libraries separately

### ISSUE THREAD

**Issue by osmoc** [kaitai-io/kaitai_struct#361]
According to readme the user is expected to "add KS runtime library for your particular language into your project". This makes KS adoption particularly hard for packages already available in distro repositories (Debian, Fedora etc) because of the additional dependency currently missing from those repositories.

Having runtime libraries packaged and available via distribution repositories would greatly simplify this step.

**Comment by GreyCat** [comment]
This is a long question without a single simple answer.

There are currently 10 languages with 10 distinct runtime libraries. Some of the are packaged by language-specific means (Python, Ruby, Java, JavaScript), some are not yet (C#, PHP, Perl), some are impossible to package, as language lacks these means (i.e. C++ or Lua), some are probably even not practical to package, as it would/might affect performance (C++).

You're asking to publish these runtime libraries to several distro repositories ("Debian, Fedora etc"), which is just not practical for a team of our size. For starters, I believe nobody in our team has accounts to publish these, and even if it was possible, it is a lot of work to package every one of them.

We do publish our own Debian repository with ksc, and probably the best thing I can see as remotely possible is to publish *some* runtimes into it as deb packages (probably auto-converted from relevant language-specific packages).

**Comment by osmoc** [comment]
> some are impossible to package, as language lacks these means (i.e. C++)

That's not true: for example boost is well-known C++ library which is packaged for pretty-much every GNU/Linux distro

> as it would/might affect performance (C++)

Could you elaborate? How packaging library might affect performance?

On a related note - is there any KS tools we can use to quantify this? I mean some benchmarking code.

In general, I agree that it's not a simple undertaking which is also might be unfeasible for all of the supported languages. Although having some of the runtime libraries (C++, Python) available directly from distro repos would be a great first step to facilitate KS adoption.

**Comment by GreyCat** [comment]
> That's not true: for example boost is well-known C++ library which is packaged for pretty-much every GNU/Linux distro

Please read the whole sentence. I said that C++ lacks *language-specific* packaging format (i.e. like Java has jars + maven central repo, like Python has eggs + pypi, Ruby has gems + rubygems.org, C# has nuget packages + nuget.org feed).

> Could you elaborate? How packaging library might affect performance?

See #347. The basic idea is that inlining certain widely used functions would be faster than calling them every time, and inlining would be pretty much impossible with a dynamically linked shared object library.

> On a related note - is there any KS tools we can use to quantify this? I mean some benchmarking code.

This particular one is not yet implemented, as per #347. But, in general, there is [benchmarks repo](https://github.com/kaitai-io/kaitai_struct_benchmarks).

> Although having some of the runtime libraries (C++, Python) available directly from distro repos would be a great first step to facilitate KS adoption.

I'd wait for #347 to be completed for C++. Python's runtime is actually already widely available virtually everywhere: there is [pypi package](https://pypi.python.org/pypi/kaitaistruct), there is [3rd party Debian package](https://packages.debian.org/buster/python-kaitaistruct), there is [3rd party Ubuntu package](https://packages.ubuntu.com/search?keywords=python-kaitaistruct), there is [3rd party Arch Linux package](https://www.archlinux.org/packages/community/any/python-kaitaistruct/), [3rd party Mageia package](https://madb.mageia.org/package/show/source/1/application/0/release/cauldron/name/python-kaitaistruct), etc, etc.

**Comment by osmoc** [comment]
> Python's runtime is actually already widely available

Nice, it might be worth it to highlight this in readme/wiki.

**Comment by GreyCat** [comment]
> Nice, it might be worth it to highlight this in readme/wiki.

There's [Python-specific notes](http://doc.kaitai.io/lang_python.html) section in the docs, feel free to contribute some sort of installation guide / list there — https://github.com/kaitai-io/kaitai_struct_doc/blob/master/lang_python.adoc. Unfortunately, it's pretty empty compared to other languages (i.e. [Java's notes](https://github.com/kaitai-io/kaitai_struct_doc/blob/master/lang_java.adoc)) so far.
